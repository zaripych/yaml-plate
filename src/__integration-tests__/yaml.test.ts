import {
  build,
  ROOT,
  emptyDirSafe,
  spawnAndCheck,
  sortedDirectoryContents,
} from './helpers';
import { join } from 'path';
import { ensureDir, readFile } from 'fs-extra';
import yaml from 'js-yaml';

describe('yaml-plate', () => {
  const outCombinedDir = join(ROOT, 'integration-test-combine-yaml');
  const outDir = join(ROOT, 'integration-test-dir-yaml');
  const directory = join(__dirname, 'test-yaml-files');
  const context = join(__dirname, 'yaml.plate.config.js');
  const outFile = join(outCombinedDir, 'combined.yaml');

  beforeAll(async () => {
    await emptyDirSafe(outCombinedDir);
    await ensureDir(outCombinedDir);
    await emptyDirSafe(outDir);
    await ensureDir(outDir);

    await build();
  });

  describe('when combining input into a single file', () => {
    beforeAll(() => {
      spawnAndCheck(
        'node',
        [
          './lib/yaml-plate.js',
          '-c',
          context,
          '-p',
          `${directory}/**/*.yaml`,
          '-s',
          outFile,
          '--sorted',
        ],
        {
          cwd: ROOT,
        }
      );
    });

    it('should output the file', async () => {
      expect(await sortedDirectoryContents(outCombinedDir))
        .toMatchInlineSnapshot(`
          Array [
            "combined.yaml",
          ]
        `);

      expect(await readFile(outFile, 'utf8')).toMatchSnapshot();
    });

    describe('when output to directory', () => {
      beforeAll(() => {
        spawnAndCheck(
          'node',
          [
            './lib/yaml-plate.js',
            '-c',
            context,
            '-p',
            `${directory}/**/*.yaml`,
            '-d',
            outDir,
            '-b',
            'src/__integration-tests__/test-yaml-files',
            '--sorted',
          ],
          {
            cwd: ROOT,
          }
        );
      });

      it('should generate directory', async () => {
        expect(await sortedDirectoryContents(outDir)).toMatchInlineSnapshot(`
          Array [
            "double-empty.yaml",
            "empty-doc.yaml",
            "multiple-docs-with-empty.yaml",
            "multiple-docs.yaml",
            "single-doc.yaml",
          ]
        `);
      });

      it('contents of the directory and combined file should be same', async () => {
        const files = await sortedDirectoryContents(outDir);
        const contentsFromDir = await Promise.all(
          files.map(file => readFile(join(outDir, file), 'utf8'))
        );
        const contentFromCombined = await readFile(outFile, 'utf8');

        const dirLoaded = contentsFromDir.reduce((acc, item) => {
          const data = yaml.safeLoadAll(item);
          return [...acc, ...(data.length === 0 ? [null] : data)];
        }, []);
        const combinedLoaded = yaml.safeLoadAll(contentFromCombined);

        expect(combinedLoaded).toEqual(dirLoaded);
      });
    });
  });
});
