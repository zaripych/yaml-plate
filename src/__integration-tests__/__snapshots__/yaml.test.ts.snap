// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`yaml-plate when combining input into a single file should output the file 2`] = `
"docId: 7
content: Seventh document content
attributes:
  contributors:
    - Hat Man
---
---
docId: 8
content: Eight document content
contributors:
  - Boss Man
---
---
docId: 4
content: Fourth document content
attributes:
  contributors:
    - Homer
---
docId: 5
content: Fifth document content
attributes:
  contributors:
    - Beth
---
---
docId: 1
content: First document content
attributes:
  contributors:
    - Rick
---
docId: 2
content: Second document content
attributes:
  contributors:
    - Morty
---
docId: 3
content: Third document content
attributes:
  contributors:
    - Mr. President
"
`;
